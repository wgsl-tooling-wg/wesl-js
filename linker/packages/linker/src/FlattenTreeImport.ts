import {
  ImportCollection,
  ImportItem,
  ImportSegment,
  ImportStatement,
} from "./parse/ImportStatement.js";

export interface FlatImport {
  importPath: string[];
  modulePath: string[];
}

/**
 * Simplify importTree into a flattened map from import paths to module paths.
 *
 * @return map from import path (with 'as' renaming) to module Path
 */
export function flattenTreeImport(imp: ImportStatement): FlatImport[] {
  return recursiveResolve([], [], imp.segments, imp.finalSegment);

  /** recurse through segments of path, producing  */
  function recursiveResolve(
    resolvedImportPath: string[],
    resolvedExportPath: string[],
    remainingPath: ImportSegment[],
    finalSegment: ImportCollection | ImportItem,
  ): FlatImport[] {
    if (remainingPath.length > 0) {
      const [segment, ...rest] = remainingPath;
      const importPath = [...resolvedImportPath, segment.name];
      const modulePath = [...resolvedExportPath, segment.name];
      return recursiveResolve(importPath, modulePath, rest, finalSegment);
    } else if (finalSegment instanceof ImportCollection) {
      // resolve path with each element in the list
      return finalSegment.subTrees.flatMap(elem => {
        return recursiveResolve(
          resolvedImportPath,
          resolvedExportPath,
          elem.segments,
          elem.finalSegment,
        );
      });
    } else if (finalSegment instanceof ImportItem) {
      const importPath = [
        ...resolvedImportPath,
        finalSegment.as || finalSegment.name,
      ];
      const modulePath = [...resolvedExportPath, finalSegment.name];
      return [{ importPath, modulePath }];
    } else {
      console.error(finalSegment);
      throw new Error("unknown segment type", { cause: finalSegment });
    }
  }
}
