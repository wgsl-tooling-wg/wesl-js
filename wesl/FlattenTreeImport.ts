import { unreachable } from "@std/assert";
import type {
  ImportCollection,
  ImportItem,
  ImportSegment,
  ImportStatement,
} from "./AbstractElems.ts";

export interface FlatImport {
  importPath: string[];
  modulePath: string[];
}

/**
 * Simplify importTree into a flattened map from import paths to module paths.
 *
 * @return map from import path (with 'as' renaming) to module Path
 */
export function flattenTreeImport(imp: ImportStatement): FlatImport[] {
  return recursiveResolve([], [], imp.segments, imp.finalSegment);

  /** recurse through segments of path, producing  */
  function recursiveResolve(
    resolvedImportPath: string[],
    resolvedExportPath: string[],
    remainingPath: ImportSegment[],
    finalSegment: ImportCollection | ImportItem,
  ): FlatImport[] {
    if (remainingPath.length > 0) {
      const [segment, ...rest] = remainingPath;
      const importPath = [...resolvedImportPath, segment.name];
      const modulePath = [...resolvedExportPath, segment.name];
      return recursiveResolve(importPath, modulePath, rest, finalSegment);
    } else if (finalSegment.kind === "import-collection") {
      // resolve path with each element in the list
      return finalSegment.subtrees.flatMap((elem) => {
        return recursiveResolve(
          resolvedImportPath,
          resolvedExportPath,
          elem.segments,
          elem.finalSegment,
        );
      });
    } else if (finalSegment.kind === "import-item") {
      const importPath = [
        ...resolvedImportPath,
        finalSegment.as ?? finalSegment.name,
      ];
      const modulePath = [...resolvedExportPath, finalSegment.name];
      return [{ importPath, modulePath }];
    } else {
      unreachable(finalSegment);
    }
  }
}
