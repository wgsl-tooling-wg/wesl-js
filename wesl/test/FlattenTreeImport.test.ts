import type { ImportCollection, ImportStatement } from "../AbstractElems.ts";
import { flattenTreeImport } from "../FlattenTreeImport.ts";
import { assertSnapshot } from "@std/testing/snapshot";

Deno.test("complex tree import", async (t) => {
  const list: ImportCollection = {
    kind: "import-collection",
    subtrees: [
      {
        kind: "import-statement",
        segments: [],
        finalSegment: { kind: "import-item", name: "foo", as: "bar" },
      },
      {
        kind: "import-statement",
        segments: [],
        finalSegment: { kind: "import-item", name: "doh" },
      },
      {
        kind: "import-statement",
        segments: [{ kind: "import-segment", name: "bib" }],
        finalSegment: { kind: "import-item", name: "bog" },
      },
    ],
  };

  const tree: ImportStatement = {
    kind: "import-statement",
    segments: [
      {
        kind: "import-segment",
        name: "zap",
      },
    ],
    finalSegment: list,
  };
  const flattened = flattenTreeImport(tree);
  await assertSnapshot(t, flattened);
});
