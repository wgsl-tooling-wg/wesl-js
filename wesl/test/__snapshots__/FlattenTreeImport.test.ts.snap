export const snapshot = {};

snapshot[`complex tree import 1`] = `
[
  {
    importPath: [
      "zap",
      "bar",
    ],
    modulePath: [
      "zap",
      "foo",
    ],
  },
  {
    importPath: [
      "zap",
      "doh",
    ],
    modulePath: [
      "zap",
      "doh",
    ],
  },
  {
    importPath: [
      "zap",
      "bib",
      "bog",
    ],
    modulePath: [
      "zap",
      "bib",
      "bog",
    ],
  },
]
`;
